File: fabric-audit-agent/core/stagger.js
/** Add minutes to an "HH:MM" string, wrapping at 24h. Pure. */
function addMinutes(hhmm, mins) {
  const [h, m] = String(hhmm).split(':').map(Number);
  const total = ((h * 60 + m + mins) % 1440 + 1440) % 1440;
  return `${String(Math.floor(total / 60)).padStart(2, '0')}:${String(total % 60).padStart(2, '0')}`;
}

/**
 * Propose a staggered refresh schedule for colliding refresh times. The largest model
 * keeps its slot; the rest are pushed out in `spacingMin` increments. Pure.
 * @param {object} facts  reads facts.capacity.refreshes [{ workspace, dataset, scheduledAt, sizeGB }]
 * @param {{ spacingMin?:number, minGroup?:number }} [opts]
 * @returns {Array<{ dataset:string, workspace:string, from:string, to:string }>}
 */
export function planStagger(facts = {}, { spacingMin = 15, minGroup = 2 } = {}) {
  const refreshes = facts?.capacity?.refreshes ?? [];
  const byTime = {};
  for (const r of refreshes) (byTime[r.scheduledAt] ??= []).push(r);

  const plan = [];
  for (const [time, group] of Object.entries(byTime)) {
    if (group.length < minGroup) continue;
    const sorted = [...group].sort((a, b) => (b.sizeGB ?? 0) - (a.sizeGB ?? 0)); // largest first keeps slot
    sorted.forEach((r, i) => {
      const to = addMinutes(time, i * spacingMin);
      if (to !== r.scheduledAt) plan.push({ dataset: r.dataset, workspace: r.workspace, from: r.scheduledAt, to });
    });
  }
  return plan;
}

File: fabric-audit-agent-app/fabric_audit_agent/stagger.py
"""Propose a staggered refresh schedule for colliding times. Port of ``core/stagger.js``. Pure."""


def _add_minutes(hhmm, mins):
    parts = str(hhmm).split(":")
    h, m = int(parts[0]), int(parts[1])
    total = ((h * 60 + m + mins) % 1440 + 1440) % 1440
    return f"{total // 60:02d}:{total % 60:02d}"


def plan_stagger(facts=None, spacing_min=15, min_group=2):
    facts = facts or {}
    refreshes = (facts.get("capacity") or {}).get("refreshes") or []
    by_time = {}
    for r in refreshes:
        by_time.setdefault(r.get("scheduledAt"), []).append(r)

    plan = []
    for time, group in by_time.items():
        if len(group) < min_group:
            continue
        ordered = sorted(group, key=lambda r: -(r.get("sizeGB") or 0))   # largest first keeps its slot
        for i, r in enumerate(ordered):
            to = _add_minutes(time, i * spacing_min)
            if to != r.get("scheduledAt"):
                plan.append({"dataset": r.get("dataset"), "workspace": r.get("workspace"), "from": r.get("scheduledAt"), "to": to})
    return plan
